File: onnxruntime/test/providers/qnn/udo/sample/run_udo_sample.cc
// Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries.
// SPDX-License-Identifier: MIT

// Standalone C++ reference sample for running a QNN UDO (User-Defined Op).
//
// Demonstrates the MyAdd UDO (output = input + constant) on:
//   - QNN CPU backend  (fp32 model, no QDQ wrapping)
//   - QNN HTP backend  (QDQ uint8 model, DQ -> MyAdd -> Q fusion)
//
// Build (after building ORT and the MyAdd op package):
//   g++ -std=c++17 run_udo_sample.cc \
//       -I<ort_include_dir> \
//       -L<ort_lib_dir> -lonnxruntime \
//       -Wl,-rpath,<ort_lib_dir> \
//       -o run_udo_sample
//
// Usage:
//   # CPU backend
//   ./run_udo_sample cpu myadd_fp32.onnx /path/to/libMyAddOpPackage_cpu.so
//
//   # HTP backend (x86 simulator or on-device)
//   ./run_udo_sample htp myadd_qdq.onnx /path/to/libMyAddOpPackage_htp.so

#include <algorithm>
#include <cmath>
#include <cstdio>
#include <cstring>
#include <numeric>
#include <stdexcept>
#include <string>
#include <vector>

#include "onnxruntime_cxx_api.h"
#include "onnxruntime_lite_custom_op.h"

static constexpr const char* kDomain = "example";
static constexpr const char* kQnnEpName = "QNNExecutionProvider";
static constexpr float kConstant = 2.0f;
static constexpr int kElements = 32;

// ── CPU fallback kernel (provides the schema + shape-inference that ORT
//    needs to parse the model; QNN EP runs the actual kernel via the package) ──
struct MyAdd {
  MyAdd(const OrtApi* api, const OrtKernelInfo* info) {
    OrtStatus* st = api->KernelInfoGetAttribute_float(info, "constant", &constant_);
    if (st) api->ReleaseStatus(st);
  }
  Ort::Status Compute(const Ort::Custom::Tensor<float>& X,
                      Ort::Custom::Tensor<float>* Y) {
    auto shape = X.Shape();
    const float* in = X.Data();
    float* out = Y->Allocate(shape);
    for (int i = 0; i < X.NumberOfElement(); ++i)
      out[i] = in[i] + constant_;
    return Ort::Status{nullptr};
  }
  static Ort::Status InferOutputShape(Ort::ShapeInferContext& ctx) {
    ctx.SetOutputShape(0, ctx.GetInputShape(0));
    return Ort::Status{nullptr};
  }
  float constant_ = 1.0f;
};

// Register the QNN EP plugin library and append it to session options using
// the v2 plugin API (RegisterExecutionProviderLibrary + AppendExecutionProvider_V2).
static void AppendQnnEp(Ort::Env& env,
                        Ort::SessionOptions& so,
                        const std::unordered_map<std::string, std::string>& ep_options) {
  // Register the QNN EP shared library with the environment.
  env.RegisterExecutionProviderLibrary(kQnnEpName, "libonnxruntime_providers_qnn.so");

  // Query all registered EP devices and filter to those from the QNN EP.
  std::vector<Ort::ConstEpDevice> all_devices = env.GetEpDevices();
  std::vector<Ort::ConstEpDevice> qnn_devices;
  for (const auto& dev : all_devices) {
    if (std::string(dev.EpName()) == kQnnEpName)
      qnn_devices.push_back(dev);
  }
  if (qnn_devices.empty())
    throw std::runtime_error("No QNN EP device found after registration.");

  so.AppendExecutionProvider_V2(env, qnn_devices, ep_options);
}

static std::vector<float> make_input() {
  std::vector<float> v(kElements);
  for (int i = 0; i < kElements; ++i)
    v[i] = -1.0f + 2.0f * i / (kElements - 1);
  return v;
}

static void run_cpu(const std::string& model_path, const std::string& pkg_path) {
  printf("\n=== QNN CPU backend ===\n");

  Ort::Env env(ORT_LOGGING_LEVEL_WARNING, "UDOSampleCPU");
  Ort::SessionOptions so;

  // 1. Register custom-op schema so ORT can load the "example" domain model.
  Ort::CustomOpDomain domain{kDomain};
  auto op = Ort::Custom::CreateLiteCustomOp<MyAdd>("MyAdd", "CPUExecutionProvider");
  domain.Add(op);
  so.Add(domain);

  // 2. Register QNN EP and configure with op_packages.
  std::string op_packages = "MyAdd:" + pkg_path + ":MyAddOpPackageInterfaceProvider";
  AppendQnnEp(env, so, {{"backend_type", "cpu"}, {"op_packages", op_packages}});

  Ort::Session session(env, model_path.c_str(), so);

  // 3. Run inference with float32 input in [-1, 1].
  auto input = make_input();
  std::vector<int64_t> shape = {1, kElements};
  auto mem = Ort::MemoryInfo::CreateCpu(OrtArenaAllocator, OrtMemTypeDefault);
  auto in_tensor = Ort::Value::CreateTensor<float>(mem, input.data(), input.size(),
                                                   shape.data(), shape.size());
  const char* in_names[] = {"input"};
  const char* out_names[] = {"output"};
  auto outputs = session.Run(Ort::RunOptions{}, in_names, &in_tensor, 1, out_names, 1);

  // 4. Verify output ≈ input + constant.
  const float* out = outputs[0].GetTensorData<float>();
  float max_err = 0.0f;
  for (int i = 0; i < kElements; ++i)
    max_err = std::max(max_err, std::fabs(out[i] - (input[i] + kConstant)));

  printf("Max absolute error vs (input + %.1f): %.2e\n", kConstant, max_err);
  printf(max_err <= 1e-4f ? "PASS\n" : "FAIL: error exceeds threshold\n");
}

static void run_htp(const std::string& model_path, const std::string& pkg_path) {
  printf("\n=== QNN HTP backend ===\n");

  Ort::Env env(ORT_LOGGING_LEVEL_WARNING, "UDOSampleHTP");
  Ort::SessionOptions so;

  // 1. Register custom-op schema.
  Ort::CustomOpDomain domain{kDomain};
  auto op = Ort::Custom::CreateLiteCustomOp<MyAdd>("MyAdd", "CPUExecutionProvider");
  domain.Add(op);
  so.Add(domain);

  // 2. Register QNN HTP EP.
  std::string op_packages = "MyAdd:" + pkg_path + ":MyAddOpPackageInterfaceProvider:CPU";
  AppendQnnEp(env, so, {{"backend_type", "htp"},
                        {"offload_graph_io_quantization", "0"},
                        {"op_packages", op_packages}});

  Ort::Session session(env, model_path.c_str(), so);

  // 3. Run inference.
  auto input = make_input();
  std::vector<int64_t> shape = {1, kElements};
  auto mem = Ort::MemoryInfo::CreateCpu(OrtArenaAllocator, OrtMemTypeDefault);
  auto in_tensor = Ort::Value::CreateTensor<float>(mem, input.data(), input.size(),
                                                   shape.data(), shape.size());
  const char* in_names[] = {"input"};
  const char* out_names[] = {"output"};
  auto outputs = session.Run(Ort::RunOptions{}, in_names, &in_tensor, 1, out_names, 1);

  // 4. Verify within QDQ tolerance (~2 quantization steps for uint8 over [-1,1]).
  const float* out = outputs[0].GetTensorData<float>();
  float max_err = 0.0f;
  for (int i = 0; i < kElements; ++i)
    max_err = std::max(max_err, std::fabs(out[i] - (input[i] + kConstant)));

  // QDQ tolerance: 2x the output quantization scale (4/255, covering [0,4] output range).
  const float qdq_tol = 4.0f / 255.0f * 2;
  printf("Max absolute error vs (input + %.1f): %.4f  (QDQ tol: %.4f)\n",
         kConstant, max_err, qdq_tol);
  printf(max_err <= qdq_tol ? "PASS\n" : "FAIL: error exceeds QDQ tolerance\n");
}

int main(int argc, char* argv[]) {
  if (argc < 4) {
    fprintf(stderr,
            "Usage:\n"
            "  %s cpu  <model.onnx> <libMyAddOpPackage_cpu.so>\n"
            "  %s htp  <model.onnx> <libMyAddOpPackage_htp.so>\n",
            argv[0], argv[0]);
    return 1;
  }

  std::string backend = argv[1];
  std::string model   = argv[2];
  std::string pkg     = argv[3];

  try {
    if (backend == "cpu")       run_cpu(model, pkg);
    else if (backend == "htp")  run_htp(model, pkg);
    else {
      fprintf(stderr, "Unknown backend '%s'. Use 'cpu' or 'htp'.\n", backend.c_str());
      return 1;
    }
  } catch (const Ort::Exception& e) {
    fprintf(stderr, "ORT error: %s\n", e.what());
    return 1;
  } catch (const std::exception& e) {
    fprintf(stderr, "Error: %s\n", e.what());
    return 1;
  }

  return 0;
}

File: onnxruntime/test/providers/qnn/udo/sample/run_udo_sample.py
# Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries.
# SPDX-License-Identifier: MIT
"""
Python reference sample: run MyAdd UDO on QNN EP.

MyAdd computes: output = input + constant
  input  : shape [1, 32], float32 values in [-1, 1]
  output : shape [1, 32], float32

Usage:
  # QNN CPU backend (fp32 model)
  python run_udo_sample.py cpu myadd_fp32.onnx \
      --schema-lib ./libMyAddSchema.so \
      --op-package ./libMyAddOpPackage_cpu.so

  # QNN HTP backend (QDQ model, x86 simulator or on-device)
  python run_udo_sample.py htp myadd_qdq.onnx \
      --schema-lib ./libMyAddSchema.so \
      --op-package ./libMyAddOpPackage_htp.so

Why --schema-lib?
  The ONNX model contains a node in the custom domain "example".  ORT needs a
  schema for that domain just to *load* the model.  In Python the only way to
  provide this is via SessionOptions.register_custom_ops_library(), which loads
  a shared lib that exposes RegisterCustomOps().  Build register_myadd_schema.cc
  to produce libMyAddSchema.so (see README.md).  This library also serves as the
  CPU fallback; actual QNN execution comes from the --op-package library.

QNN EP registration (v2 plugin API):
  The QNN EP is registered as a plugin library via
  ort.register_execution_provider_library() + ort.get_ep_devices() +
  so.add_provider_for_devices(), matching the C++ AppendExecutionProvider_V2 path.
"""

import argparse
import os
import sys
import numpy as np
import onnxruntime as ort

CONSTANT = 2.0
INPUT_SHAPE = (1, 32)
QNN_EP_NAME = "QNNExecutionProvider"


def build_input() -> np.ndarray:
    n = INPUT_SHAPE[1]
    return np.linspace(-1.0, 1.0, n, dtype=np.float32).reshape(INPUT_SHAPE)


def append_qnn_ep(so: ort.SessionOptions, qnn_ep_lib: str, ep_options: dict) -> None:
    """Register the QNN EP plugin and append it to session options (v2 API).

    qnn_ep_lib must be an absolute path to libonnxruntime_providers_qnn.so.
    """
    ort.register_execution_provider_library(QNN_EP_NAME, qnn_ep_lib)
    devices = [d for d in ort.get_ep_devices() if d.ep_name == QNN_EP_NAME]
    if not devices:
        raise RuntimeError("No QNN EP device found after registration.")
    so.add_provider_for_devices(devices, ep_options)


def run_cpu(model_path: str, schema_lib: str, op_package: str, qnn_ep_lib: str) -> None:
    print("\n=== QNN CPU backend ===")
    so = ort.SessionOptions()
    so.register_custom_ops_library(schema_lib)

    op_packages_str = f"MyAdd:{op_package}:MyAddOpPackageInterfaceProvider"
    append_qnn_ep(so, qnn_ep_lib, {"backend_type": "cpu", "op_packages": op_packages_str})

    sess = ort.InferenceSession(model_path, sess_options=so)
    x = build_input()
    [output] = sess.run(["output"], {"input": x})

    expected = x + CONSTANT
    max_err = float(np.max(np.abs(output - expected)))
    print(f"Max absolute error vs (input + {CONSTANT}): {max_err:.2e}")
    if max_err > 1e-4:
        print("FAIL: error exceeds threshold")
        sys.exit(1)
    print("PASS")


def run_htp(model_path: str, schema_lib: str, op_package: str, qnn_ep_lib: str) -> None:
    print("\n=== QNN HTP backend ===")
    so = ort.SessionOptions()
    so.register_custom_ops_library(schema_lib)

    op_packages_str = f"MyAdd:{op_package}:MyAddOpPackageInterfaceProvider:CPU"
    append_qnn_ep(so, qnn_ep_lib, {
        "backend_type": "htp",
        "offload_graph_io_quantization": "0",
        "op_packages": op_packages_str,
    })

    sess = ort.InferenceSession(model_path, sess_options=so)
    x = build_input()
    [output] = sess.run(["output"], {"input": x})

    expected = x + CONSTANT
    max_err = float(np.max(np.abs(output - expected)))
    # QDQ tolerance: 2x the output quantization scale (4/255, covering [0,4] output range).
    qdq_tol = 4.0 / 255.0 * 2
    print(f"Max absolute error vs (input + {CONSTANT}): {max_err:.4f}  (QDQ tol: {qdq_tol:.4f})")
    if max_err > qdq_tol:
        print("FAIL: error exceeds QDQ tolerance")
        sys.exit(1)
    print("PASS")


def main() -> None:
    parser = argparse.ArgumentParser(description="Run MyAdd UDO on QNN EP")
    parser.add_argument("backend", choices=["cpu", "htp"], help="QNN backend to use")
    parser.add_argument("model", help="Path to ONNX model (myadd_fp32.onnx or myadd_qdq.onnx)")
    parser.add_argument("--schema-lib", required=True,
                        help="Path to libMyAddSchema.so (built from register_myadd_schema.cc)")
    parser.add_argument("--op-package", required=True,
                        help="Path to libMyAddOpPackage_<backend>.so")
    parser.add_argument("--qnn-ep-lib", required=True,
                        help="Absolute path to libonnxruntime_providers_qnn.so")
    args = parser.parse_args()

    if args.backend == "cpu":
        run_cpu(args.model, args.schema_lib, args.op_package, args.qnn_ep_lib)
    else:
        run_htp(args.model, args.schema_lib, args.op_package, args.qnn_ep_lib)


if __name__ == "__main__":
    main()
